### PR TITLE
[#611] Added 'CacheTrait' with targeted Drupal cache invalidation steps

### LINKED ISSUE #611: Add 'CacheTrait' with targeted Drupal cache invalidation steps.
# Add `CacheTrait` for targeted Drupal cache invalidation steps

## Summary

Add a new `Drupal/CacheTrait` with steps for invalidating specific caches
— a single page by path, a path pattern, or the render cache — from
inside a Behat scenario.

## Motivation

Scenarios that verify cache behaviour (cache-tag invalidation, Max-Age
headers, render cache misses, aggregator feeds) need precise cache
invalidation, not a blanket clear. `DrupalContext` already provides
`Given the cache has been cleared` for the global case, but there is no
step-level API for targeted invalidation. Targeted clears are also much
faster than a full rebuild and let scenarios prove that a specific tag
actually invalidates what it should.

This issue adds three targeted cache-clearing steps — a single path, a
wildcard path pattern, and the render cache. A full clear is
deliberately out of scope because it is already provided upstream.

## Proposed steps

```gherkin
Given the page cache for the path :path has been cleared
Given the page cache for the paths matching :path_pattern has been cleared
Given the render cache has been cleared
```

### Step definitions

| Step | Method |
|---|---|
| `Given the page cache for the path :path has been cleared` | `cacheClearPagePath` |
| `Given the page cache for the paths matching :path_pattern has been cleared` | `cacheClearPagePathWildcard` |
| `Given the render cache has been cleared` | `cacheClearRender` |

Phrasing matches `DrupalContext::@Given the cache has been cleared` (past participle state, not "is cleared" active voice).

## Implementation notes

- New file: `src/Drupal/CacheTrait.php`.
- `cacheClearPagePath($path)` — invalidate by URL cache tag for a single
  path. Use `Cache::invalidateTags(['http_response', 'url:' . $path])` or
  the `page_cache` cache bin directly.
- `cacheClearPagePathWildcard($pattern)` — convert the glob-style pattern
  to a SQL `LIKE` query against the `cache_page` table (or the
  internal_page_cache backend), then delete matching entries.
- `cacheClearRender()` — `\Drupal::cache('render')->deleteAll()`.
- All steps are `Given` because they are setup, not actions under test.
- Method names: `cacheClearPagePath`, `cacheClearPagePathWildcard`,
  `cacheClearRender`.

### Configurable defaults (protected methods)

```php
// The cache bin used for page cache — override if the site uses a custom
// internal page cache bin name.
protected function cacheGetPageCacheBin(): string {
  return 'page';
}
```

## Scope boundaries

- No per-cache-tag invalidation step — tags are an implementation detail
  that scenarios should not care about directly.
- No "clear all caches" step — `DrupalContext::@Given the cache has been
  cleared` already covers the global case.
- No `drush cr` wrapper — DrupalExtension's `DrushContext` already
  provides `I run drush :command`.

## Acceptance criteria

- [ ] New trait in `src/Drupal/CacheTrait.php`.
- [ ] Steps use `#[Given(...)]` tuple annotations.
- [ ] Wildcard pattern conversion safely escapes SQL `LIKE` metacharacters
      other than the wildcard itself.
- [ ] Feature tests in `tests/behat/features/drupal_cache.feature`.
- [ ] `@trait:CacheTrait` negative tests for invalid paths.
- [ ] `cacheGetPageCacheBin()` protected helper.
- [ ] `ahoy update-docs` run.
- [ ] `ahoy lint` passes.
- [ ] Add `use CacheTrait;` to `FeatureContext.php`.

### PR DESCRIPTION
**Pull request by AlexSkrypnyk**
## Summary

Adds a new `Drupal\CacheTrait` with three targeted cache-clearing `Given` steps for scenarios that need precise invalidation without a full site-wide rebuild:

- `Given the page cache for the path :path has been cleared` — invalidates the `url:<path>` and `http_response` cache tags for a single URL.
- `Given the page cache for the paths matching :path_pattern has been cleared` — deletes matching rows from the `cache_page` bin, converting glob `*` to SQL `%` and escaping the other `LIKE` metacharacters (`\`, `%`, `_`).
- `Given the render cache has been cleared` — calls `\Drupal::cache('render')->deleteAll()`.

The trait exposes a `cacheGetPageCacheBin()` protected helper so sites with a custom internal page cache bin can override the default `page` bin name. A blanket cache clear is deliberately out of scope — `DrupalContext::@Given the cache has been cleared` already covers it.

Fixes #611

## Acceptance checklist

- [x] New trait in `src/Drupal/CacheTrait.php`.
- [x] Steps use `#[Given(...)]` tuple annotations.
- [x] Wildcard pattern conversion safely escapes SQL `LIKE` metacharacters other than the wildcard itself.
- [x] Feature tests in `tests/behat/features/drupal_cache.feature`.
- [x] `@trait:Drupal\CacheTrait` negative tests for invalid paths.
- [x] `cacheGetPageCacheBin()` protected helper.
- [x] `use CacheTrait;` added to `tests/behat/bootstrap/FeatureContext.php`.
- [ ] `ahoy update-docs` run.
- [ ] `ahoy lint` passes.

## Test plan

- [ ] Run `ahoy test-bdd tests/behat/features/drupal_cache.feature` and verify all positive and `@trait` negative scenarios pass.
- [ ] Run `ahoy lint` to confirm PHPStan, Rector, PHPCS, and Gherkinlint are clean.
- [ ] Run `ahoy update-docs` and confirm `STEPS.md` includes the three new steps.

## Summary

This PR introduces a new `Drupal\CacheTrait` providing three cache invalidation steps for Behat scenarios: clearing page cache for a single path, clearing page cache for paths matching a glob-style pattern, and clearing the render cache.

## Changes

- **src/Drupal/CacheTrait.php**: New trait with three public step methods (`cacheClearPagePath()`, `cacheClearPagePathWildcard()`, `cacheClearRender()`) and a protected override hook (`cacheGetPageCacheBin()`). Includes input validation, SQL escaping for wildcard patterns, and direct database operations for the wildcard matcher.
- **tests/behat/bootstrap/FeatureContext.php**: Added trait usage `use DrevOps\BehatSteps\Drupal\CacheTrait;`
- **tests/behat/features/drupal_cache.feature**: New feature file with 7 scenarios covering success cases and negative test cases for invalid inputs (empty paths, missing leading slashes).
- **README.md**: Added entry to the Drupal steps index linking to `Drupal\CacheTrait`.
- **STEPS.md**: Added documentation section with all three step definitions and usage examples.

## Critical Issues

**Step Definition Type Violation**: The three cache-clearing steps are incorrectly defined as `@Given` steps but are used as `@When` steps in the feature file. Per CONTRIBUTING.md:
- `Given` defines test prerequisites and should use words like "exists" or "have"
- `When` describes actions with action verbs
- Steps describing actions (like "has been cleared") are actions, not prerequisites, and must be `When` steps
- The standard pattern for action steps is `When I <verb>`

The steps are currently defined with `#[Given(...)]` annotations but used as:
- `When the page cache for the path "/user" has been cleared` (line 9 of feature file)
- `When the page cache for the paths matching "/user*" has been cleared` (line 15)
- `When the render cache has been cleared` (line 21)

These must be corrected to use `#[When(...)]` annotations with the proper "I" format (e.g., `When I clear the page cache for the path :path`) to comply with the project's step definition standards.

## Other Observations

- Input validation is thorough (empty path checks, leading slash validation)
- SQL LIKE metacharacter escaping logic is correct
- Negative test coverage for invalid inputs is comprehensive
- Documentation has been properly generated and added

### CODE REVIEW

File: STEPS.md [hunk 1]
```diff
@@ -2611,6 +2612,59 @@ Then the block "My block" should not exist in the "content" region
...
+  <summary><code>@Given the page cache for the paths matching :path_pattern has been cleared</code></summary>
+
+<br/>
+Clear the page cache for all paths matching a glob-style pattern
+<br/><br/>
+
+```gherkin
+Given the page cache for the paths matching "/news" has been cleared
+
+```
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply STEPS.md:2650]: Fixed in 3ac0218. The docs generator strips slash-star from docblock lines (docs.php:341), so /news/star was being truncated in STEPS.md. Changed the example to /news* which round-trips cleanly and still shows the wildcard syntax.

File: tests/behat/features/drupal_cache.feature [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  Scenario: Assert "Given the page cache for the paths matching :path_pattern has been cleared" clears matching paths
+    Given I am logged in as a user with the "administrator" role
+    When the page cache for the paths matching "/user*" has been cleared
+    Then I go to "/user"
+
+  @api
+  Scenario: Assert "Given the render cache has been cleared" clears the render cache
+    Given I am logged in as a user with the "administrator" role
+    When the render cache has been cleared
+    Then I go to "/user"
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply tests/behat/features/drupal_cache.feature]: Fixed in 3ac0218. Moved the cache-clearing prerequisites to And (matching Given semantics), promoted the navigation to When, and added a Then assertion for each of the three positive scenarios.
